fix: Add validation for incompatible Datadog and trace context propagation

abernix opened the pull request:
Prevents simultaneous activation of Datadog tracing and W3C trace context propagation to avoid trace ID format conflicts.

<!-- start metadata -->

<!-- [ROUTER-####] -->
---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] PR description explains the motivation for the change and relevant context for reviewing
- [x] PR description links appropriate GitHub/Jira tickets (creating when necessary)
- [x] Changeset is included for user-facing changes
- [x] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [x] Performance impact assessed and acceptable
- [x] Metrics and logs are added[^3] and documented
- Tests added and passing[^4]
    - [x] Unit tests
    - [x] Integration tests, as necessary
    - [x] Manual tests, as necessary

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: A lot of (if not most) features benefit from built-in observability and `debug`-level logs. Please read [this guidance](https://github.com/apollographql/router/blob/dev/dev-docs/metrics.md#adding-new-metrics) on metrics best-practices.
[^4]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

abernix committed:
fix: Add validation for incompatible Datadog and trace context propagation

Prevents simultaneous activation of Datadog tracing and W3C trace context propagation to avoid trace ID format conflicts.

abernix commented on apollo-router/src/plugins/telemetry/mod.rs:
```diff
@@ -1117,6 +1120,26 @@ impl Telemetry {
+
+        // Check if both Datadog and trace context propagation are active
+        let datadog_active = propagation.datadog || tracing.datadog.enabled();
+        let trace_context_active = propagation.trace_context || tracing.otlp.enabled;
```
2: Is also in any similar way true that `tracing.otlp` by default will implicitly enable the `trace_context` propagator?

abernix commented on apollo-router/src/plugins/telemetry/mod.rs:
```diff
@@ -1117,6 +1120,26 @@ impl Telemetry {
+        let tracing = &config.exporters.tracing;
+
+        // Check if both Datadog and trace context propagation are active
+        let datadog_active = propagation.datadog || tracing.datadog.enabled();
```
1: Is it in any way true that `datadog.enabled()` by itself will turn on `datadog` propagation?

goto-bus-stop replied:
Yes, though it might be nice to not duplicate this condition, which is also used in `Telemetry::create_propagator`. Or put a comment on here and inside `create_propagator` that we must keep them in sync

abernix committed:
refactor: DRY up propagation activation logic with helper functions

Extract duplicated logic for determining if Datadog and trace_context
propagation are active into helper functions:

- datadog.enabled() implicitly enables Datadog propagation
- otlp.enabled implicitly enables trace context propagation

carodewig commented on apollo-router/src/plugins/telemetry/mod.rs:
```diff
@@ -1117,6 +1120,41 @@ impl Telemetry {
         TextMapCompositePropagator::new(propagators)
     }
 
+    /// Check if Datadog propagation is active.
```
Nit: could these be methods on the `Tracing` struct instead? Then you could do `config.exporters.tracing.is_datadog_propagation_active()`, etc

abernix commented:
Putting this back to draft, because many of the integration tests need some work before they'll pass.

BrynCooke commented:
I'm taking this over
